File: scanpy_neighbors/critic.py
"""
scanpy_neighbors critic — Evaluates kNN graph quality

FEEDBACK KEYS (must match parameter_science_guide conditions in skill.yaml):
    fragmented_clusters    → n_neighbors too low; graph under-connected
    over_merged_clusters   → n_neighbors too high; neighborhoods too large
    noisy_structure        → n_pcs too high; noise in distance calculation
    missing_rare_populations → n_pcs too low; rare variation excluded
"""
import numpy as np


def critic_post_process(adata, context=None):
    context = context or {}
    protocol = context.get('protocol', '10x Genomics')

    thresholds = {
        '10x Genomics': {'min_neighbors': 5,  'max_neighbors': 50, 'min_conn': 0.01, 'max_conn': 0.5},
        'Smart-seq2':   {'min_neighbors': 10, 'max_neighbors': 50, 'min_conn': 0.02, 'max_conn': 0.6},
        'Drop-seq':     {'min_neighbors': 5,  'max_neighbors': 40, 'min_conn': 0.01, 'max_conn': 0.5},
        'CITE-seq':     {'min_neighbors': 10, 'max_neighbors': 50, 'min_conn': 0.01, 'max_conn': 0.5},
    }
    t = thresholds.get(protocol, thresholds['10x Genomics'])

    metrics = {}
    warnings = []
    feedback = []

    if 'neighbors' not in adata.uns or 'connectivities' not in adata.obsp:
        return {
            'metrics': {},
            'warnings': ['CRITICAL: neighbors not computed — run sc.pp.neighbors() first'],
            'feedback': [],
            'success': False,
            'context_used': protocol
        }

    params_used = adata.uns['neighbors'].get('params', {})
    n_neighbors = params_used.get('n_neighbors', 0)
    conn = adata.obsp['connectivities']
    mean_conn = float(conn.data.mean()) if conn.nnz > 0 else 0.0

    metrics['n_neighbors_used'] = n_neighbors
    metrics['mean_connectivities'] = mean_conn
    metrics['graph_density'] = float(conn.nnz / (adata.n_obs ** 2))

    if n_neighbors < t['min_neighbors']:
        warnings.append(
            f"UNDER-CONNECTED GRAPH: n_neighbors={n_neighbors} is very low "
            f"(min {t['min_neighbors']} for {protocol}). Graph may be fragmented."
        )
        feedback.append('fragmented_clusters')

    if n_neighbors > t['max_neighbors']:
        warnings.append(
            f"OVER-CONNECTED GRAPH: n_neighbors={n_neighbors} is high "
            f"(max {t['max_neighbors']} for {protocol}). May over-merge distinct populations."
        )
        feedback.append('over_merged_clusters')

    if mean_conn < t['min_conn']:
        warnings.append(
            f"WEAK CONNECTIVITY: mean connectivity={mean_conn:.4f} is very low. "
            "Graph may not reflect biological similarity well."
        )
        feedback.append('noisy_structure')

    return {
        'metrics': metrics,
        'warnings': warnings,
        'feedback': feedback,
        'success': not bool(feedback),
        'context_used': protocol
    }

File: seurat_findneighbors/critic.R
#' seurat_findneighbors critic — Evaluates SNN graph quality
#'
#' FEEDBACK KEYS (match parameter_science_guide in skill.yaml):
#'   missing_rare_populations → dims too few; rare variation excluded
#'   noisy_clusters           → dims too many; noise in distance calc
#'   fragmented_clusters      → k.param too low; graph under-connected
#'   over_merged_populations  → k.param too high; populations merged

critic_post_process <- function(obj, context = NULL) {
    if (is.null(context)) context <- list()
    protocol <- if (!is.null(context$protocol)) context$protocol else "10x Genomics"

    thresholds <- list(
        "10x Genomics" = list(min_dims = 5L,  max_dims = 50L, min_k = 5L,  max_k = 50L),
        "Smart-seq2"   = list(min_dims = 10L, max_dims = 50L, min_k = 10L, max_k = 50L),
        "Drop-seq"     = list(min_dims = 5L,  max_dims = 40L, min_k = 5L,  max_k = 40L),
        "CITE-seq"     = list(min_dims = 10L, max_dims = 50L, min_k = 10L, max_k = 50L)
    )
    t <- if (!is.null(thresholds[[protocol]])) thresholds[[protocol]] else thresholds[["10x Genomics"]]

    metrics  <- list()
    warnings <- character(0)
    feedback <- character(0)

    # Check graph exists
    graph_names <- names(obj@graphs)
    snn_found <- any(grepl("snn", tolower(graph_names)))
    if (!snn_found) {
        return(list(
            metrics      = list(),
            warnings     = "CRITICAL: No SNN graph found — run FindNeighbors() first",
            feedback     = character(0),
            success      = FALSE,
            context_used = protocol
        ))
    }

    # Extract params from analysis_history if available
    history <- obj@misc$analysis_history
    n_dims_used  <- NA_integer_
    k_param_used <- NA_integer_

    if (!is.null(history) && length(history) > 0) {
        for (entry in rev(history)) {
            if (!is.null(entry$skill_id) && entry$skill_id == "seurat_findneighbors") {
                n_dims_used  <- entry$metrics$n_dims_used
                k_param_used <- entry$metrics$k_param_used
                break
            }
        }
    }

    metrics$n_cells      <- as.integer(ncol(obj))
    metrics$n_dims_used  <- n_dims_used
    metrics$k_param_used <- k_param_used

    if (!is.na(n_dims_used)) {
        if (n_dims_used < t$min_dims) {
            warnings <- c(warnings, sprintf(
                "TOO FEW DIMS: %d dims used (min %d for %s). Rare populations may be missed.",
                n_dims_used, t$min_dims, protocol
            ))
            feedback <- c(feedback, "missing_rare_populations")
        }
        if (n_dims_used > t$max_dims) {
            warnings <- c(warnings, sprintf(
                "TOO MANY DIMS: %d dims used (max %d for %s). Noise may dominate graph.",
                n_dims_used, t$max_dims, protocol
            ))
            feedback <- c(feedback, "noisy_clusters")
        }
    }

    if (!is.na(k_param_used)) {
        if (k_param_used < t$min_k) {
            warnings <- c(warnings, sprintf(
                "LOW k.param: %d (min %d for %s). Graph may be fragmented.",
                k_param_used, t$min_k, protocol
            ))
            feedback <- c(feedback, "fragmented_clusters")
        }
        if (k_param_used > t$max_k) {
            warnings <- c(warnings, sprintf(
                "HIGH k.param: %d (max %d for %s). Distinct populations may be merged.",
                k_param_used, t$max_k, protocol
            ))
            feedback <- c(feedback, "over_merged_populations")
        }
    }

    list(
        metrics      = metrics,
        warnings     = warnings,
        feedback     = feedback,
        success      = length(feedback) == 0L,
        context_used = protocol
    )
}
